src/test/clojure/clojure/tools/gitlibs/impl_test.cljc

```diff
@@ -32,3 +39,197 @@
 
     ;; git - unknown transport with url rewrite in gitconfig (unusual, but do something useful)
     "work:repo.git" "ssh/work/repo"))
+
+;; Test helpers for concurrent clone coordination
+
+(defn- delete-dir-recursive
+  "Recursively deletes a directory and all its contents."
+  [^DirectoryInfo dir]
+  (when (.Exists dir)
+    (Directory/Delete (.FullName dir) true)))
+
+(defn- delete-file-if-exists
+  "Deletes a file if it exists."
+  [^String path]
+  (when (File/Exists path)
+    (File/Delete path)))
+
+(defn- run-ensure-git-dir-process
+  "Runs ensure-git-dir in a separate process. Returns {:proc Process :output-file String}."
+  [url output-file]
+  (let [proc-info (doto (ProcessStartInfo. "cljr" (str "-M -m clojure.tools.gitlibs.impl " url))
+                    (.set_RedirectStandardOutput true)
+                    (.set_RedirectStandardError true)
+                    (.set_UseShellExecute false))
+        proc (Process/Start proc-info)]
+    {:proc proc :output-file output-file}))
+
+(defn- wait-for-process
+  "Waits for process to complete and returns {:exit int :stdout String :stderr String}."
+  [^Process proc]
+  (.WaitForExit proc)
+  {:exit (.ExitCode proc)
+   :stdout (let [stream (.StandardOutput proc)] (.ReadToEnd stream))
+   :stderr (let [stream (.StandardError proc)] (.ReadToEnd stream))})
+
+(defn- get-lockfile
+  "Returns the lockfile path for a given URL."
+  [url]
+  (let [git-dir-file (impl/git-dir url)]
+    (#'impl/lockfile-for-git-dir git-dir-file)))
+
+(defn- get-config-file-path
+  "Returns the config file path for a given URL."
+  [url]
+  (let [git-dir-file (impl/git-dir url)
+        config-file (cio/file-info git-dir-file "config")]
+    (.FullName config-file)))
+
+;; Scenario 1: We won the race (we created the lock)
+;; - Thread should realize it lost and must wait
+;; - If we delete lockfile when git dir is deleted, thread should throw
+;; - If we delete lockfile when git dir exists, thread should return normally
```

Comment:
Add one more test to scenario 1: if we stop updating the lockfile, the other thread should throw after 10 seconds.